aten/src/ATen/CPUApplyUtils.h

```diff
@@ -22,289 +27,348 @@ namespace at {
-        << " to have the same number of elements, but got " << t1_numel << ", " << t2_numel << ", and " << t3_numel << " elements respectively";
-    throw std::runtime_error(oss.str());
+  oss << "and " << tensors[tensors.size() - 1]
+      << " to have the same number of elements, but got ";
+  for (size_t i = 0; i < tensors.size() - 1; i++) {
+    oss << tensors[i].numel() << ", ";
   }
+  oss << "and " << tensors[tensors.size() - 1].numel()
+      << " elements respectively";
+  return oss.str();
+}
 
-  while(!TH_TENSOR_APPLY_hasFinished)
-  {
-    /* Loop through the inner most region of the Tensor */
-    for(; tensor1_i <  tensor1_size && tensor2_i < tensor2_size && tensor3_i < tensor3_size; tensor1_i++, tensor2_i++, tensor3_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride, tensor3_data += tensor3_stride)
-    {
-      op(*tensor1_data, *tensor2_data, *tensor3_data);
+AT_API inline bool _apply_preamble(ArrayRef<Tensor> tensors) {
+  checkBackend("CPU_tensor_apply", tensors, Backend::CPU);
+  if (!_all_equal_numel(tensors))
+    throw std::runtime_error(_all_equal_numel_error(tensors));
+  // An empty tensor has no elements
+  for (auto& t : tensors)
+    if (t.sizes().equals({0}))
+      return false;
+  internal::init_tbb_num_threads();
+  return true;
+}
+
+AT_API inline void iterate_all(){};
+
+template <typename Arg, typename... Args>
+AT_API inline void iterate_all(Arg& iter, Args&... iter_tail) {
+  iterate(iter);
+  iterate_all(iter_tail...);
+}
+
+AT_API inline void forward_all(int64_t offset){};
+
+template <typename Arg, typename... Args>
+AT_API inline void forward_all(int64_t offset, Arg& iter, Args&... iter_tail) {
+  forward(offset, iter);
+  forward_all(offset, iter_tail...);
+}
+
+AT_API inline bool continue_loop() {
+  return true;
+}
+
+template <typename Arg, typename... Args>
+AT_API inline bool continue_loop(Arg& iter, Args&... iter_tail) {
+  return iter.counter_[iter.dim_ - 1] < iter.sizes_[iter.dim_ - 1] &&
+      continue_loop(iter_tail...);
+}
+
+AT_API inline int64_t max_dim() {
+  return 0;
+}
+
+template <typename Arg, typename... Args>
+AT_API inline int64_t max_dim(Arg& iter, Args&... iter_tail) {
+  return std::max(iter.dim_, max_dim(iter_tail...));
+}
+
+AT_API inline void increase_counters(){};
+
+template <typename Arg, typename... Args>
+AT_API inline void increase_counters(Arg& iter, Args&... iter_tail) {
+  iter.counter_[iter.dim_ - 1]++;
+  iter.data_ += iter.strides_[iter.dim_ - 1];
+  increase_counters(iter_tail...);
+}
+
+AT_API inline void apply_op(){};
+
+template <typename Op, typename... Args>
+AT_API inline void
+apply_op(int64_t numel, int64_t offset, const Op& op, Args... iters) {
+  // For 0-dim tensors
+  if (numel == 1 && max_dim(iters...) == 0) {
+    op(*iters.data_...);
+    return;
+  }
+  forward_all(offset, iters...);
+  for (int64_t i = 0; i < numel;) {
+    for (; continue_loop(iters...) && i < numel;
```

Comment:
Can `continue_loop` ever become `false` before `i` reaches `numel`?

Reply:
Sorry I didn't get that. We do assert that the numel is matching, which means that no matter how many counters every tensor has, they should all wrap at the same time, exactly when `numel` is reached. What am I missing again?